sklearn/metrics/cluster/unsupervised.py

```diff
@@ -180,22 +182,24 @@ def silhouette_samples(X, labels, metric='euclidean', **kwds):
     # cluster in inter_clust_dists[i]
     inter_clust_dists = np.inf * intra_clust_dists
 
-    for curr_label in unique_labels:
+    for curr_label in range(len(unique_labels)):
 
         # Find inter_clust_dist for all samples belonging to the same
         # label.
         mask = labels == curr_label
         current_distances = distances[mask]
 
         # Leave out current sample.
-        n_samples_curr_lab = np.sum(mask) - 1
+        n_samples_curr_lab = n_samples_per_label[curr_label] - 1
         if n_samples_curr_lab != 0:
             intra_clust_dists[mask] = np.sum(
                 current_distances[:, mask], axis=1) / n_samples_curr_lab
+        else:
+            intra_clust_dists[mask] = 0
```

Comment:
Is it worthy to initialize `intra_clust_dists` with np.zeros this at the start and get rid of the else?